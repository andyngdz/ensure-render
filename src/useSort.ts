import { TSortValues, IUseSort, ICurrentSort } from './types//sort'
import { useDeepCompareEffect } from 'react-use'
import { useState, useEffect } from 'react'
import produce from 'immer'

const useSort = (fields: Array<string>, initSort: ICurrentSort): IUseSort => {
  const [current, setCurrent] = useState<ICurrentSort>(initSort)
  const [sorts, setSorts] = useState<TSortValues>()

  useDeepCompareEffect(() => {
    const { field: initField, direction } = initSort
    const newSorts: TSortValues = {}

    fields.forEach(field => {
      newSorts[field] = {
        active: initField === field,
        direction: initField === field ? direction : 'ASC'
      }
    })

    setSorts(newSorts)
  }, [fields])

  useEffect(() => {
    if (sorts) {
      let current

      Object.keys(sorts).forEach(field => {
        const { active, direction } = sorts[field]
        if (active) {
          current = { field, direction }
        }
      })

      if (current) setCurrent(current)
    }
  }, [sorts])

  const resetSortActive = (): void => {
    setSorts(prevState =>
      produce(prevState, draft => {
        if (draft && sorts) {
          Object.keys(sorts).forEach(field => {
            draft[field].active = false
          })
        }
      })
    )
  }

  const onSort = (field: string) => {
    resetSortActive()

    setSorts(prevState =>
      produce(prevState, draft => {
        if (sorts && draft) {
          const { direction } = sorts[field]

          if (field !== current.field) {
            draft[field].direction = 'ASC'
          } else {
            draft[field].direction = direction === 'ASC' ? 'DESC' : 'ASC'
          }

          draft[field].active = true
        }
      })
    )
  }

  return { current, sorts, onSort }
}

export { useSort }
