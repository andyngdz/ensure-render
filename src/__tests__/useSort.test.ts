import { renderHook, act } from '@testing-library/react-hooks'
import { useSort } from '../useSort'

describe('useSort', () => {
  it('Should init correctly', () => {
    const {
      result: { current }
    } = renderHook(() =>
      useSort(['id', 'name'], {
        field: 'id',
        direction: 'DESC'
      })
    )

    expect(current).toMatchSnapshot()
  })

  it('Should return state correctly after calling onSort', async () => {
    const { result } = renderHook(() =>
      useSort(['id', 'name'], {
        field: 'id',
        direction: 'DESC'
      })
    )

    await act(async () => {
      result.current.onSort('name')
    })

    expect(result.current).toMatchSnapshot()
  })
})
