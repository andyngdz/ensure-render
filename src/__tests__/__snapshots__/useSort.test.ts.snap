// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useSort Should init correctly 1`] = `
Object {
  "current": Object {
    "direction": "DESC",
    "field": "id",
  },
  "onSort": [Function],
  "sorts": Object {
    "id": Object {
      "active": true,
      "direction": "DESC",
    },
    "name": Object {
      "active": false,
      "direction": "DESC",
    },
  },
}
`;

exports[`useSort Should return state correctly after calling onSort 1`] = `
Object {
  "current": Object {
    "direction": "ASC",
    "field": "name",
  },
  "onSort": [Function],
  "sorts": Object {
    "id": Object {
      "active": false,
      "direction": "DESC",
    },
    "name": Object {
      "active": true,
      "direction": "ASC",
    },
  },
}
`;
